datalake_api/querier.py

```diff
@@ -210,20 +210,22 @@ def query_by_time(self, start, end, what, where=None, cursor=None):
             buckets = buckets[i:]
 
         for b in buckets:
-            kwargs = self._prepare_time_bucket_kwargs(b, what,
-                                                      limit=MAX_RESULTS/2)
-            if where is not None:
-                self._add_range_key_condition(kwargs, where)
-            if cursor is not None:
-                self._add_cursor_conditions(kwargs, cursor)
-
-            response = self._table.query(**kwargs)
-            results += self._exclude_outside(response['Items'], start, end)
-            # we _could_ deduplicate the results here to make more headroom for
-            # another bucket.
-            cursor = self._cursor_for_time_query(response, results, b)
-            if cursor is not None:
-                break
+            while len(results) <= MAX_RESULTS/2:
+                kwargs = self._prepare_time_bucket_kwargs(b, what,
+                                                          limit=MAX_RESULTS/2)
+                if where is not None:
+                    self._add_range_key_condition(kwargs, where)
+                if cursor is not None:
+                    self._add_cursor_conditions(kwargs, cursor)
+
+                response = self._table.query(**kwargs)
+                results += self._exclude_outside(response['Items'], start, end)
+                # we _could_ deduplicate the results here to make more headroom
+                # for another bucket.
+                cursor = self._cursor_for_time_query(response, results, b)
+                if cursor is None:
+                    # no more results in the bucket
+                    break
```

Comment:
Ok yeah, that was my original plan, but I remember suggesting it and getting shot down, though I don't recall specifically.